pkg/cloudprovider/provider/linode/provider.go

```diff
@@ -0,0 +1,447 @@
+		"label": machine.Spec.Name,
+	})
+
+	listOptions := linodego.NewListOptions(0, string(filter))
+	linodes, err := client.ListInstances(ctx, listOptions)
+
+	if err != nil {
+		return nil, doStatusAndErrToTerminalError(err)
+	}
+
+	for i, linode := range linodes {
+		if linode.Label == machine.Spec.Name && sets.NewString(linode.Tags...).Has(string(machine.UID)) {
+			return &linodeInstance{linode: &linodes[i]}, nil
+		}
+	}
+
+	return nil, cloudprovidererrors.ErrInstanceNotFound
+}
+
+func (p *provider) MigrateUID(machine *v1alpha1.Machine, new types.UID) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	c, _, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return fmt.Errorf("failed to decode providerconfig: %v", err)
+	}
+	client := getClient(c.Token)
+	linodes, err := client.ListInstances(ctx, nil)
+	if err != nil {
+		return fmt.Errorf("failed to list linodes: %v", err)
+	}
+
+	for _, linode := range linodes {
+		if linode.Label == machine.Spec.Name && sets.NewString(linode.Tags...).Has(string(machine.UID)) {
+			updateOpts := linode.GetUpdateOptions()
+
+			tags := []string{string(new)}
+			if updateOpts.Tags != nil {
+				oldUID := string(machine.UID)
+				for _, existingTag := range *updateOpts.Tags {
+					if existingTag != oldUID {
+						tags = append(tags, existingTag)
+					}
+				}
+			}
+			updateOpts.Tags = &tags
+			_, err = client.UpdateInstance(ctx, linode.ID, updateOpts)
+			if err != nil {
+				return fmt.Errorf("failed to revise linode UID tags: %v", err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (p *provider) GetCloudConfig(spec v1alpha1.MachineSpec) (config string, name string, err error) {
+	return "", "", nil
+}
+
+func (p *provider) MachineMetricsLabels(machine *v1alpha1.Machine) (map[string]string, error) {
+	labels := make(map[string]string)
+
+	c, _, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err == nil {
+		labels["type"] = c.Type
+		labels["region"] = c.Region
+	}
+
+	return labels, err
+}
+
+type linodeInstance struct {
+	linode *linodego.Instance
+}
+
+func (d *linodeInstance) Name() string {
+	return d.linode.Label
+}
+
+func (d *linodeInstance) ID() string {
+	return strconv.Itoa(d.linode.ID)
+}
+
+func (d *linodeInstance) Addresses() []string {
+	var addresses []string
+	for _, n := range d.linode.IPv4 {
+		addresses = append(addresses, n.String())
+	}
+	addresses = append(addresses, d.linode.IPv6)
+	return addresses
+}
+
+func (d *linodeInstance) Status() instance.Status {
+	switch d.linode.Status {
+	case "new":
+		return instance.StatusCreating
+	case "active":
+		return instance.StatusRunning
+	default:
+		return instance.StatusUnknown
+	}
+}
+
+// doStatusAndErrToTerminalError judges if the given HTTP status
+// can be qualified as a "terminal" error, for more info see v1alpha1.MachineStatus
+
+// if the given error doesn't qualify the error passed as
+// an argument will be returned
+func doStatusAndErrToTerminalError(err error) error {
```

Comment:
Can we rename this to `linodeStatusAndErroToTerminalError`? :)

Reply:
Sure .. I wasn't sure if "do" here was a verb or a reference to DO.

Reply:
These status strings will be updated to reflect linodego constants.

Reply:
updated in 575403a